lib/sycamore/sycamore/utils/pdf.py

```diff
@@ -0,0 +1,115 @@
+import logging
+
+from io import BytesIO
+from PIL import Image
+from queue import Queue
+from subprocess import PIPE, Popen
+from threading import Thread
+from typing import List, Generator
+
+from sycamore.utils.time_trace import LogTime
+
+
+def convert_from_path_streamed(pdf_path: str) -> Generator[Image.Image, None, None]:
+    class StdoutEOF:
+        pass
+
+    class StderrEOF:
+        pass
+
+    def capture_exception(q, fn, finish_msg):
+        try:
+            fn()
+        except e:
+            q.put(e)
+
+        q.put(finish_msg)
+
+    def read_stdout(fh, q):
+        HEADER_BYTES = 40
+        need_bytes = HEADER_BYTES
+        data = b""
+        while True:
+            if need_bytes > len(data):
+                logging.debug(f"reading. have {len(data)}/{need_bytes}")
+                part = fh.read(need_bytes - len(data))
+                if part == b"":  # Eof
+                    assert len(data) == 0  # nothing left
+                    break
+                data = data + part
+            else:
+                logging.debug(f"no reading. have {len(data)}/{need_bytes}")
+
+            if len(data) < need_bytes:
+                continue
+
+            code, size, rgb = tuple(data[0:HEADER_BYTES].split(b"\n")[0:3])
+            size_x, size_y = tuple(size.split(b" "))
+            file_size = len(code) + len(size) + len(rgb) + 3 + int(size_x) * int(size_y) * 3
+
+            if len(data) < file_size:
+                need_bytes = file_size
+                continue
+
+            img = Image.open(BytesIO(data[0:file_size])).convert("RGB")
+            q.put(img)
+            data = data[file_size:]
+            need_bytes = HEADER_BYTES
+
+    def read_stderr(fh, q):
+        while True:
+            line = fh.readline()
+            if line == b"":
+                break
+
+            q.put(str(line, encoding="utf-8").rstrip())
+
+    with LogTime("convert_to_image"):
+        # If we don't do this, then if the stderr buffer fills up we could get stuck.
+        # Popen.communicate() reads the entire strings.
+        args = ["pdftoppm", "-r", "200", pdf_path]
+        proc = Popen(args, stdout=PIPE, stderr=PIPE)
+        q: Queue = Queue()
+        t_out = Thread(target=capture_exception, args=(q, lambda: read_stdout(proc.stdout, q), StdoutEOF()))
```

Comment:
What joins or otherwise cleans up these threads?

Reply:
I believe they're auto-cleaned up when the thread exits (which has to have happened for the loop to finish), but I added the explicit joins.